virl2_client/models/auth_management.py

```diff
@@ -0,0 +1,391 @@
+
+    def __init__(self, session: Client, auto_sync=True, auto_sync_interval=1.0):
+        self.auto_sync = auto_sync
+        self.auto_sync_interval = auto_sync_interval
+        self._last_sync_time = 0.0
+        self._session = session
+        self._method = AuthMethod.LOCAL
+        self._managers: dict[AuthMethod, Optional[AuthMethodManager]] = {
+            AuthMethod("local"): None,
+            AuthMethod("ldap"): LDAPManager(self),
+        }
+
+    def sync_if_outdated(self) -> None:
+        timestamp = time.time()
+        if (
+            self.auto_sync
+            and timestamp - self._last_sync_time > self.auto_sync_interval
+        ):
+            self.sync()
+
+    def sync(self) -> None:
+        settings: dict = self._session.get(self._METHOD_URL).json()
+        self._method = AuthMethod(settings.pop("method"))
+        self._last_sync_time = time.time()
+        # Since right now we also get the other settings along with the method, we
+        # might as well save a request and store said settings in the relevant manager
+        if (manager := self._managers[self._method]) is not None:
+            manager._settings = settings
+            manager._last_sync_time = time.time()
+
+    @property
+    def method(self) -> AuthMethod:
+        self.sync_if_outdated()
+        return self._method
+
+    @method.setter
+    def method(self, value: AuthMethod | str):
+        value = AuthMethod(value)
+        self._session.put(self._METHOD_URL, json={"method": value.value})
+        self._method = value
+
+    @property
+    def manager(self) -> AuthMethodManager:
+        """
+        The manager for the current auth method.
+        """
+        # TODO: should we allow users to also configure currently inactive auth methods?
+        #  not sure how useful that might or might not be, but if they can then that
+        #  changes the answer to the question in AuthMethodManager.update_setting
+        self.sync_if_outdated()
+        return self._managers[self._method]
+
+
+class AuthMethodManager:
+    _METHOD = ""
+    _CONFIG_URL = _BASE_URL + "/config"
+    _TEST_URL = _BASE_URL + "/test"
+
+    def __init__(self, auth_management: AuthManagement):
+        self._auth_management = auth_management
+        self._last_sync_time = 0.0
+        self._settings = {}
+
+    def sync_if_outdated(self) -> None:
+        timestamp = time.time()
+        if (
+            self._auth_management.auto_sync
+            and timestamp - self._last_sync_time
+            > self._auth_management.auto_sync_interval
+        ):
+            self.sync()
+
+    def sync(self) -> None:
+        self._settings = self._get_current_settings()
+        self._last_sync_time = time.time()
+
+    @property
+    def _session(self) -> Client:
+        return self._auth_management._session
+
+    def _get_current_settings(self) -> dict:
+        settings = self._session.get(self._CONFIG_URL).json()
+        del settings["method"]
```

Comment:
why do we delete method?  I don't think it's a big deal.

Reply:
Redundancy; you are supposed to get method from AuthManagement, not a manager. Also, if we split it away from the settings later, that would remove it from here anyway, so I'm just doing it ahead of time so that users don't expect it to stay.